src/platform_impl/android/keycodes.rs

```diff
@@ -156,324 +159,393 @@ pub fn to_physical_keycode(keycode: Keycode) -> KeyCode {
     }
 }
 
-// TODO: We need to expose getUnicodeChar via android-activity instead of having
-// a fixed mapping from key codes
-pub fn to_logical(keycode: Keycode, native: NativeKey) -> Key {
-    use android_activity::input::Keycode::*;
+/// Tries to map the `key_event` to a `KeyMapChar` containing a unicode character or dead key accent
+///
+/// This takes a `KeyEvent` and looks up its corresponding `KeyCharacterMap` and
+/// uses that to try and map the `key_code` + `meta_state` to a unicode
+/// character or a dead key that can be combined with the next key press.
+pub fn character_map_and_combine_key(
+    app: &AndroidApp,
+    key_event: &KeyEvent<'_>,
+    combining_accent: &mut Option<char>,
+) -> Option<KeyMapChar> {
+    let device_id = key_event.device_id();
+
+    let key_map = match app.device_key_character_map(device_id) {
+        Ok(key_map) => key_map,
+        Err(err) => {
+            log::error!("Failed to look up `KeyCharacterMap` for device {device_id}: {err:?}");
+            return None;
+        }
+    };
+
+    match key_map.get(key_event.key_code(), key_event.meta_state()) {
+        Ok(KeyMapChar::Unicode(unicode)) => {
+            // Only do dead key combining on key down
+            if key_event.action() == KeyAction::Down {
+                let combined_unicode = if let Some(accent) = combining_accent {
+                    match key_map.get_dead_key(*accent, unicode) {
```

Comment:
I think you can combine both `if let` and `match` with the help of `map`/`then`/`mar_or`.

Reply:
It doesn't really seem like it to me. The outer if let can't be done with a map since it's mapping the Some and None both to Some.

Maybe there's some elaborate combinator version of this code but it's not standing out for me and I think it's reasonable enough as it is.